flink-connectors/flink-connector-gcp-pubsub/src/main/java/org/apache/flink/streaming/connectors/gcp/pubsub/PubSubSink.java

```diff
@@ -273,7 +290,13 @@ private PubSubSinkBuilder(SerializationSchema<IN> serializationSchema) {
 		 */
 		public PubSubSink<IN> build() throws IOException {
 			if (credentials == null) {
-				credentials = defaultCredentialsProviderBuilder().build().getCredentials();
+				if (hostAndPort == null) {
+					// No hostAndPort is the normal scenario so we use the default credentials.
+					credentials = defaultCredentialsProviderBuilder().build().getCredentials();
+				} else {
+					// With hostAndPort the PubSub emulator is used so we do not have credentials.
+					credentials = EmulatorCredentials.getInstance();
+				}
```

Comment:
Again, I guess credentials handling should be done outside of the connector (so that we can pass different credentials for the test)

Reply:
You can already pass different credentials if you want to via the withCredentials method in both the normal and the emulator situations.

This code is about which are the right **default** credentials to choose in case none were provided: Either the default Google credentials or (in case the emulator was specified) the "Emulator Credentials".